### ISSUE TITLE
mlpack no longer compiles on RHEL 7 / CentOS 7

### ISSUE THREAD

**Issue by conradsnicta** [mlpack/mlpack#2939]
Tried to compile mlpack master (as of 07 May 2021) on a standard RHEL 7 / CentOS 7 box and got this:

```
(...)
-- Could NOT find Boost: Found unsuitable version "1.53.0", but required is at least "1.58" (found /usr/include)
-- [download 0% complete]
-- [download 1% complete]
-- [download 2% complete]
(...)
-- Found Boost: /home/blah/git/mlpack-master/build/deps/boost_1_76_0/ (found version "1.76.0")  
(...)
In file included from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/config.hpp:61:0,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/variant/detail/config.hpp:16,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/variant/variant_fwd.hpp:16,
                 from /home/blah/git/mlpack-master/build/deps/cereal-1.3.0/include/cereal/types/boost_variant.hpp:41,
                 from /home/blah/git/mlpack-master/src/mlpack/prereqs.hpp:91,
                 from /home/blah/git/mlpack-master/build/src/mlpack/cotire/mlpack_CXX_prefix.cxx:4,
                 from /home/blah/git/mlpack-master/build/src/mlpack/cotire/mlpack_CXX_prefix.hxx:4:
/home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/precision.hpp:41:83: error: redeclaration ‘int boost::math::tools::digits()’ differs in ‘constexpr’
 inline BOOST_MATH_CONSTEXPR int digits(BOOST_MATH_EXPLICIT_TEMPLATE_TYPE_SPEC(T)) BOOST_NOEXCEPT
                                                                                   ^
In file included from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/precision.hpp:16:0,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/fraction.hpp:13,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/special_functions/gamma.hpp:19,
                 from /home/blah/git/mlpack-master/src/mlpack/core/kernels/spherical_kernel.hpp:13,
                 from /home/blah/git/mlpack-master/src/mlpack/core.hpp:109,
                 from /home/blah/git/mlpack-master/src/mlpack/bindings/tests/clean_memory.cpp:14:
/home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/policies/policy.hpp:25:15: error: from previous declaration ‘constexpr int boost::math::tools::digits()’
 constexpr int digits(BOOST_MATH_EXPLICIT_TEMPLATE_TYPE(T)) noexcept;
(...)
In file included from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/series.hpp:16:0,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/special_functions/gamma.hpp:18,
                 from /home/blah/git/mlpack-master/src/mlpack/core/kernels/spherical_kernel.hpp:13,
                 from /home/blah/git/mlpack-master/src/mlpack/core.hpp:109,
                 from /home/blah/git/mlpack-master/src/mlpack/bindings/cli/print_help.hpp:15,
                 from /home/blah/git/mlpack-master/src/mlpack/bindings/cli/print_help.cpp:13:
/home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/config.hpp:261:77: error: redeclaration ‘T boost::math::tools::epsilon()’ differs in ‘constexpr’
 #  define BOOST_MATH_NOEXCEPT(T) noexcept(boost::is_floating_point<T>::value)
                                                                             ^
/home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/precision.hpp:246:82: note: in expansion of macro ‘BOOST_MATH_NOEXCEPT’
 inline BOOST_MATH_CONSTEXPR T epsilon(BOOST_MATH_EXPLICIT_TEMPLATE_TYPE_SPEC(T)) BOOST_MATH_NOEXCEPT(T)
                                                                                  ^
In file included from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/precision.hpp:16:0,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/tools/fraction.hpp:13,
                 from /home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/special_functions/gamma.hpp:19,
                 from /home/blah/git/mlpack-master/src/mlpack/core/kernels/spherical_kernel.hpp:13,
                 from /home/blah/git/mlpack-master/src/mlpack/core.hpp:109,
                 from /home/blah/git/mlpack-master/src/mlpack/bindings/cli/print_help.hpp:15,
                 from /home/blah/git/mlpack-master/src/mlpack/bindings/cli/print_help.cpp:13:
/home/blah/git/mlpack-master/build/deps/boost_1_76_0/boost/math/policies/policy.hpp:27:13: error: from previous declaration ‘constexpr T boost::math::tools::epsilon()’
 constexpr T epsilon(BOOST_MATH_EXPLICIT_TEMPLATE_TYPE(T)) noexcept(std::is_floating_point<T>::value);
(...)
```

RHEL 7 / CentOS 7 has gcc 4.8 as the default compiler.
```
$ g++ --version
g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
```

**Comment by shrit** [comment]
Ah great, happy to see that the autodownloader have worked out of the box in this case, it did find an old non-compatible version and redownloaded the latest one.
The problem is related to boost and gcc, as it seems that it is not able to compile boost. In my humble option, the best solution here is to accelerate removing boost::math rather than find a fix for RHEL 7.
Another quick solution, is to choose the best working boost version and hosted on mlpack website and download it from there until we remove boost completely. I am not sure what version we need to choose in this case.

**Comment by conradsnicta** [comment]
> Another quick solution, is to choose the best working boost version and hosted on mlpack website and download it from there until we remove boost completely. I am not sure what version we need to choose in this case.

I suggest the above solution.  Probably take the lowest version that mlpack can handle.

RHEL 7 / CentOS 7 is still used in a lot of corporate and managed academic environments, so it'd be advantageous to keep mlpack working in such environments.

PS. Boost touts itself as being peer "reviewed", but I'm not entirely convinced.  The error messages shown by gcc 4.8 suggest that there is a problem with the BOOST_MATH_CONSTEXPR macro, which may be conditionally defined.

This is not the first time that updating Boost has broken stuff.  I got burned a while back and refuse to use Boost.  

Last time I checked Boost didn't have a policy of not breaking stuff from one minor version to another.  They certainly don't follow [semantic versioning](https://semver.org/). It's more of a playground where ideas for new C++ standards are tried out, reformulated, changed, etc.

**Comment by shrit** [comment]
I will open a pull request for this one, I will download the lowest version, and see if we can host it on mlpack website since I can not trust the old version links, they can break easily.
We do have boost 1.76 in the matrix but not with the same compiler, maybe it is a good idea to add gcc 4.8 or not I am not sure, 4.8 is still very old

**Comment by rcurtin** [comment]
I can't wait until we don't have a Boost dependency anymore! :)

There's not really a way to get a newer compiler than gcc 4.8 on RHEL7.  I know that EPEL7, however, has the `boost169` package, which can be used here.  @conradsnicta you could try installing that package, and I believe then configuration/building of mlpack should work fine.

@shrit do you think we should add a CMake check to see if the compiler is gcc 4.x, and if so, we would use the autodownloader to download an older version of Boost?  Personally speaking, I think when building on most systems, it is better to try and avoid the autodownloader unless it's totally necessary---in fact, I think our Boost autodownloader has no way to actually install Boost once downloaded, so unless you are building a static version of mlpack, there may be later problems.  Maybe we should update the documentation to suggest that users don't use the autodownloader?  (Of course, this Boost installation issue becomes no longer an issue when we no longer have a Boost dependency. :))

**Comment by shrit** [comment]
If you agree with this, I will open a pull request to host the older version on mlpack website (1.58), otherwise, they will definitely change the link.
There will be no need for boost installation because the master version does not require any installation process.
I will open another one to test the gcc compiler and download the older version in this case.

**Comment by birm** [comment]
In trying to test #2940, I got the following:

```
[ 23%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/cf/cf_model.cpp.o
In file included from /mlpack/src/mlpack/core/tree/binary_space_tree/binary_space_tree.hpp:573:0,
                 from /mlpack/src/mlpack/core/tree/binary_space_tree.hpp:23,
                 from /mlpack/src/mlpack/methods/neighbor_search/neighbor_search.hpp:20,
                 from /mlpack/src/mlpack/methods/cf/cf.hpp:20,
                 from /mlpack/src/mlpack/methods/cf/cf_model.hpp:17,
                 from /mlpack/src/mlpack/methods/cf/cf_model.cpp:12:
/mlpack/src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp: In substitution of 'template<bool B, class T> using enable_if_t = typename std::enable_if::type [with bool B = cereal::is_loading<arma::Mat<double> >(0u); T = void]':
/mlpack/src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp:554:1:   required by substitution of 'template<class Archive> mlpack::tree::BinarySpaceTree<MetricType, StatisticType, MatType, BoundType, SplitType>::BinarySpaceTree(Archive&, std::enable_if_t<is_loading<Archive>()>*) [with Archive = arma::Mat<double>]'
/mlpack/src/mlpack/methods/cf/neighbor_search_policies/cosine_search.hpp:58:50:   required from here
/mlpack/src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp:554:1: internal compiler error: Segmentation fault
 BinarySpaceTree<MetricType, StatisticType, MatType, BoundType, SplitType>::
 ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
Preprocessed source stored into /tmp/cc9wsl7e.out file, please attach this to your bugreport.
make[2]: *** [src/mlpack/CMakeFiles/mlpack.dir/methods/cf/cf_model.cpp.o] Error 1
make[1]: *** [src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
make: *** [all] Error 2
```

Still looking to see if something else may be wrong with my setup, but figured someone would be interested.

**Comment by conradsnicta** [comment]
> /mlpack/src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp:554:1: internal compiler error: Segmentation fault

@birm Is this with gcc 4.8 on CentOS 7 / RHEL 7, or a more recent compiler?

If the former, it's unlikely to be fixed by GCC and/or RH.  A workaround in mlpack would be required.

RHEL 7 is still supported by RH, but it's in the "Maintenance Support 2" phase, where only mission critical stuff is fixed (ie. mainly security fixes).
* https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#Product_life_cycle
* https://access.redhat.com/support/policy/updates/errata/

**Comment by rcurtin** [comment]
@birm same here:

```
/root/mlpack/src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:390:1:   required by substitution of 'template<class Archive> mlpack::tree::RectangleTree<MetricType, StatisticType, MatType, SplitType, DescentType, AuxiliaryInformationType>::RectangleTree(Archive&, std::enable_if_t<is_loading<Archive>()>*) [with Archive = mlpack::tree::RectangleTree<mlpack::metric::LMetric<2, true>, mlpack::kde::KDEStat, arma::Mat<double>, mlpack::tree::RTreeSplit, mlpack::tree::RTreeDescentHeuristic, mlpack::tree::NoAuxiliaryInformation>]'
/root/mlpack/src/mlpack/methods/kde/kde_impl.hpp:117:21:   required from 'mlpack::kde::KDE<KernelType, MetricType, MatType, TreeType, DualTreeTraversalType, SingleTreeTraversalType>::KDE(const mlpack::kde::KDE<KernelType, MetricType, MatType, TreeType, DualTreeTraversalType, SingleTreeTraversalType>&) [with KernelType = mlpack::kernel::TriangularKernel; MetricType = mlpack::metric::LMetric<2, true>; MatType = arma::Mat<double>; TreeType = mlpack::tree::RTree; DualTreeTraversalType = mlpack::tree::RectangleTree<mlpack::metric::LMetric<2, true>, mlpack::kde::KDEStat, arma::Mat<double>, mlpack::tree::RTreeSplit, mlpack::tree::RTreeDescentHeuristic, mlpack::tree::NoAuxiliaryInformation>::DualTreeTraverser; SingleTreeTraversalType = mlpack::tree::RectangleTree<mlpack::metric::LMetric<2, true>, mlpack::kde::KDEStat, arma::Mat<double>, mlpack::tree::RTreeSplit, mlpack::tree::RTreeDescentHeuristic, mlpack::tree::NoAuxiliaryInformation>::SingleTreeTraverser]'
/root/mlpack/src/mlpack/methods/kde/kde_model.hpp:138:7:   required from 'mlpack::kde::KDEWrapper<KernelType, TreeType>* mlpack::kde::KDEWrapper<KernelType, TreeType>::Clone() const [with KernelType = mlpack::kernel::TriangularKernel; TreeType = mlpack::tree::RTree]'
/root/mlpack/src/mlpack/methods/kde/kde_model.cpp:317:1:   required from here
/root/mlpack/src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:390:1: internal compiler error: Segmentation fault
 RectangleTree<MetricType, StatisticType, MatType, SplitType, DescentType,
 ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
```

While I really like the idea of preserving compatibility of RHEL7 and in the past I would have made time for it, at least right now I just don't think I can make the time for it.  I suppose that we can at least say that mlpack 3.4.2 builds for RHEL7 (it's available in EPEL), but probably mlpack 4 won't unless someone wants to make the time to find a workaround.

**Comment by conradsnicta** [comment]
> we can at least say that mlpack 3.4.2 builds for RHEL7 (it's available in EPEL), but probably mlpack 4 won't unless someone wants to make the time to find a workaround.

Entirely reasonable given the constraints.  Any RHEL7/CentOS 7 users can always step up.

This has an upside: dropping RHEL7 also allows bumping the minimum standard of C++ to C++14.

mlpack currently requires Armadillo 8.400 (outdated at this point), which is present on Ubuntu 18.04.  In turn, Ubuntu 18.04 has gcc 7, which has the C++14 standard as default.  RHEL 8 uses gcc 8.
https://gcc.gnu.org/projects/cxx-status.html#cxx14

I'd go as far as recommending to make Ubuntu 20.04 and RHEL 8 as the minimum supported environments.  I don't think anybody who seriously uses Linux on the desktop would still be using Ubuntu 18.04 at this point.

RHEL 8 has Armadillo 9.700, while Ubuntu 20.04 has Armadillo 9.800.  Both are massive improvements over Armadillo 8.400, especially in the degree of available functionality.

**Comment by conradsnicta** [comment]
GCC 8 is available for RHEL 7 via:
* Developer Toolset 8 - https://www.softwarecollections.org/en/scls/rhscl/devtoolset-8/
* Red Hat Developer Toolset - https://access.redhat.com/documentation/en-us/red_hat_developer_toolset/8/

In other words, users are not limited to GCC 4.8 on RHEL 7, and have options to use more recent compilers.

I recommend placing a note on the mlpack site and/or README to state that at least GCC 6 or 7 is required, and add the above links for RHEL 7 users.

**Comment by rcurtin** [comment]
Agreed.  I think that would be a good way to solve things.  I opened #2955 to make those changes. :+1:

**Comment by conradsnicta** [comment]
Resolved by #2955
